### PR TITLE
Remove unstable `macro_reexport`

### LINKED ISSUE #29638: Tracking issue for `macro_reexport` feature
Tracking for stabilization of `#[macro_reexport]`.

### LINKED ISSUE #38951: Src link for re-exported macros are missing.
Consider <https://doc.rust-lang.org/core/macro.write.html> which has a [Src] link, but <https://doc.rust-lang.org/std/macro.write.html> does not.  The Std docs typically have higher visibility as it's what's linked from the docs.rust-lang.org homepage.

### PR DESCRIPTION
**Pull request by petrochenkov**
It's subsumed by `feature(use_extern_macros)` and `pub use`

cc https://github.com/rust-lang/rust/issues/35896
closes https://github.com/rust-lang/rust/issues/29638
closes https://github.com/rust-lang/rust/issues/38951

### CODE REVIEW

**Review by TimNN — COMMENTED** [review]
@petrochenkov: As far as I can tell with my limited knowledge of compiler internals, the changes look good.

Can you ping someone from the compiler team who is more familiar with this area of code, to give this a second look (otherwise I'll ping a random person from the team)?

**Comment by kennytm** [comment]
Actual failure:

```
[01:09:17] failures:
[01:09:17] 
[01:09:17] ---- [run-pass] run-pass-fulldeps/proc-macro/use-reexport.rs stdout ----
[01:09:17] 	
[01:09:17] error: aux-build `/checkout/src/test/run-pass-fulldeps/proc-macro/auxiliary/derive-reexport.rs` source not found
[01:09:17] thread '[run-pass] run-pass-fulldeps/proc-macro/use-reexport.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:1877:9
[01:09:17] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[01:09:17] 
[01:09:17] 
[01:09:17] failures:
[01:09:17]     [run-pass] run-pass-fulldeps/proc-macro/use-reexport.rs
```

File: src/libstd/lib.rs [hunk 1]
```diff
@@ -343,15 +343,13 @@ use prelude::v1::*;
...
-                 debug_assert_ne, unreachable, unimplemented, write, writeln, try)]
-extern crate core as __core;
+// Re-export a few macros from core
+#[stable(feature = "rust1", since = "1.0.0")]
+pub use core::{assert_eq, assert_ne, debug_assert, debug_assert_eq, debug_assert_ne};
+#[stable(feature = "rust1", since = "1.0.0")]
+pub use core::{unreachable, unimplemented, write, writeln, try};
 
 #[macro_use]
-#[macro_reexport(vec, format)]
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/libstd/lib.rs]: Exactly.

File: src/libstd/lib.rs [hunk 2]
```diff
@@ -310,6 +309,7 @@
 #![feature(unboxed_closures)]
 #![feature(untagged_unions)]
 #![feature(unwind_attributes)]
+#![feature(use_extern_macros)]
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/libstd/lib.rs]: I'd say unlikely.
If all local users of `libstd` like compiler and test suite build succesfully, then everything should be okay.

**Comment by petrochenkov** [comment]
cc @rust-lang/lang 
r? someone pursuing stabilization of `use_extern_macros`, e. g. @withoutboats

File: src/test/ui/macro-reexport-removed.stderr [hunk 3]
```diff
@@ -0,0 +1,18 @@
+error[E0557]: feature has been removed
+  --> $DIR/macro-reexport-removed.rs:13:12
+   |
+LL | #![feature(macro_reexport)] //~ ERROR feature has been removed
```
Comment (petrochenkov) [inline src/test/ui/macro-reexport-removed.stderr]: This should probably have some note about why the feature was removed and what should be used instead.

**Comment by pietroalbini** [comment]
Ping from triage @withoutboats! Can you review this PR?

**Comment by cramertj** [comment]
@pietroalbini I believe they're out this week.

r? @alexcrichton maybe?

**Comment by alexcrichton** [comment]
Thanks @petrochenkov! r=me once this is ready

**Comment by pietroalbini** [comment]
Seems like a spurious failure?

@bors retry

**Comment by kennytm** [comment]
Legit rustdoc test failure.

```
failures:
---- [rustdoc] rustdoc\cross-crate-links.rs stdout ----
	
error: htmldocck failed!
status: exit code: 1
command: "C:\\Python27\\python2.7.exe" "C:\\projects\\rust\\src/etc/htmldocck.py" "C:\\projects\\rust\\build\\i686-pc-windows-gnu\\test\\rustdoc\\cross-crate-links.stage2-i686-pc-windows-gnu" "C:\\projects\\rust\\src/test\\rustdoc\\cross-crate-links.rs"
stdout:
------------------------------------------
------------------------------------------
stderr:
------------------------------------------
69: @has check failed
	`XPATH PATTERN` did not match
	// @has 'foo/index.html' '//a[@href="../all_item_types/macro.foo_macro.html"]' 'foo_macro'
Encountered 1 errors
------------------------------------------
thread '[rustdoc] rustdoc\cross-crate-links.rs' panicked at 'explicit panic', tools\compiletest\src\runtest.rs:2965:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by petrochenkov** [comment]
Oh, come on.
@bors retry

**Pull request merged**